asdf/tags/core/ndarray.py

```diff
@@ -221,6 +222,7 @@ def __init__(self, source, shape, dtype, offset, strides,
         self._asdffile = asdffile
         self._source = source
         self._block = None
+        self._block_data_ref = None
```

Comment:
It might be nice to name this, indicating it is (or will be) a `weakref`.

Reply:
I'm fairly sure I would not vote to remove update. It can be an essential function if someone is trying to update a small part of an existing file. Something I might consider is if an update requires rewriting most or all blocks, just write out a new copy of the file and delete the old. That might simplify the code. An option is just to make an existing block unused and add a new one at the end if it is being replaced. That grows the file with dead space, so it would be an option someone might explicitly choose in using update. Anyway, these are topics for a different PR...

Reply:
Make a consideration also to how we might use the update in the AWS S3 environment, where update of the object store that represents the file is not allowed.. I don't believe S3 supports full CRUD.

I'm thinking here of the case for an exploded file where the blocks are stored externally. If we want to replace a block with updated data, and also need to update the reference to the block. It would be nice to update just the block, and the reference in the header instead of replacing and rewriting the entire file and it's associated blocks

Reply:
@sosey I think you are basically saying that that situation doesn't support update. For example, if you use exploded files and need to update a block, you need to rewrite the whole block (right?). In that sense, update is only on the ensemble, but not on any file. Would some sort of exploded update tool serve to do that just as well?

Reply:
If we are updating just one block, which exists externally, with a reference to it's location in the tree, I want to make sure that only the one block we are changing gets updated, and the reference to it is updated in the tree. 

I think to do this in S3 we need to delete the external block, write the new block, then delete the header object and write the new header object with everything the same except the new reference to the updated block.  

I think it's possible to have a mix of block storage types, what if data changes from internal to external. I want to not only have the case where a reference to an external array changes (because the data changed), but also the case where data that was stored internally is moved externally, for that case, it might make sense to optimize the internal array storage while updating the reference to the external data.

Reply:
I think the current state of affairs is that asdf rewrites every external block on update.  That happens here: https://github.com/spacetelescope/asdf/blob/master/asdf/block.py#L354

S3 doesn't permit updates to a range within an object, but you can overwrite an entire existing object in one operation, so it wouldn't be necessary to delete the old blocks/headers first.